[clang-tidy][readability-container-contains] Extend to any class with contains

nicovank opened the pull request:
This check will now work out of the box with other containers that have a
`contains` method, such as `folly::F14` or Abseil containers.

It will also work with strings, which are basically just weird containers.
`std::string` and `std::string_view` will have a `contains` method starting with
C++23. `llvm::StringRef` and `folly::StringPiece` are examples of existing
implementations with a `contains` method.

nicovank commented:
Pinging @vogelsgesang, @PiotrZSL, @5chmidti, @HerrCai0907, @carlosgalvezp for review.

I applied this modified version to llvm-project root, **554 files changed**, `check-all` passes. Almost all cases are moving from `count` to `contains`. How do you suggest putting up this PR for easier review? Split it up by top-level directory?

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -116,7 +116,11 @@ Changes in existing checks
-- Improved :doc:`readablility-implicit-bool-conversion
+- Improved :doc:`readability-container-contains
+  <clang-tidy/checks/readability/container-contains>` check
+  to let it work on any class that has a `contains` method.
```
```suggestion
  to let it work on any class that has a ``contains`` method.
```

Language construct.

EugeneZelenko commented on clang-tools-extra/clang-tidy/readability/ContainerContainsCheck.cpp:
```diff
@@ -15,26 +15,25 @@ using namespace clang::ast_matchers;
-                     "::std::unordered_map", "::std::multiset",
-                     "::std::unordered_multiset", "::std::multimap",
-                     "::std::unordered_multimap"))))));
+  const auto HasContainsMethod = hasMethod(cxxMethodDecl(hasName("contains")));
```
Should `contain` signature be checked?

5chmidti reviewed (commented):
> I applied this modified version to llvm-project root

One PR per subproject seems to be the preferred way, judging by the responses to some of those *all of llvm* PRs. IMO, this is a readability cleanup and would probably be accepted.

EugeneZelenko commented on clang-tools-extra/clang-tidy/readability/ContainerContainsCheck.cpp:
```diff
@@ -15,26 +15,26 @@ using namespace clang::ast_matchers;
-                     "::std::unordered_map", "::std::multiset",
-                     "::std::unordered_multiset", "::std::multimap",
-                     "::std::unordered_multimap"))))));
+  const auto HasContainsMethod = hasMethod(
+      cxxMethodDecl(isConst(), returns(booleanType()), hasName("contains")));
+  const auto ContainerWithContains = hasType(
+      hasUnqualifiedDesugaredType(recordType(hasDeclaration(cxxRecordDecl(anyOf(
+          HasContainsMethod, hasAnyBase(hasType(hasCanonicalType(hasDeclaration(
+                                 cxxRecordDecl(HasContainsMethod)))))))))));
```
And return type is `bool`.

SimplyDanny replied:
… and is sufficiently visible. This depends on the scope though, that is `private` is fine as long as the call is in another method of the type. It's not, however, if the call would require a publicly visible `contains` to be valid.

nicovank commented:
Update, now getting the same hits on llvm-project with/without type checking (553/553), a good indication no true positives are missed.

nicovank commented:
I was over-complicating the task. Instead of trying to match the type of the parameter with the type of the argument, I can just match the types of both parameters. Since on "normal" containers methods will have the same signature minus name, this shouldn't remove any true positives due to conversions.

See updated simplified PR. Number of matches on llvm-project is the same with or without `hasParameter(0, ...)` clauses.

nicovank committed:
[clang-tidy][readability-container-contains] Extend to any class with contains

This check will now work out of the box with other containers that have a
`contains` method, such as `folly::F14` or Abseil containers.

It will also work with strings, which are basically just weird containers.
`std::string` and `std::string_view` will have a `contains` method starting with
C++23. `llvm::StringRef` and `folly::StringPiece` are examples of existing
implementations with a `contains` method.

HerrCai0907 reviewed (approved):
LGTM
